util/_dask.py

```diff
@@ -0,0 +1,400 @@
+"""
+Example usage:
+
+    `Map` and `Reduce` functions for Dask distributed client.
+
+    >>> from util._dask import dask_map, dask_reduce
+
+    # map and keep track of future <-> workitem mapping
+    >>> futures, futurekey2item = dask_map(
+    >>>     Processor().process,
+    >>>     workitems,
+    >>>     client=client,
+    >>>     NanoEventsFactory_kwargs={
+    >>>         "preload": lambda b: b.name in {'Jet_pt', 'Jet_eta'},
+    >>>         "schemaclass": NtupleSchema,
+    >>>     }
+    >>> )
+
+    # perform reduction and track failures
+    >>> final_future, failed_items = dask_reduce(
+    >>>     futures,
+    >>>     futurekey2item=futurekey2item,
+    >>>     client=client,
+    >>>     treereduction=16,
+    >>> )
+
+See also: `if __name__ == "__main__":` block for a complete example.
+"""
+
+from __future__ import annotations
+
+import dataclasses
+import typing as tp
+from collections import Counter, defaultdict
+from functools import partial
+
+import awkward as ak
+import uproot
+from coffea.nanoevents import NanoEventsFactory
+from coffea.processor import Accumulatable, accumulate
+from coffea.processor.executor import WorkItem
+from coffea.util import rich_bar
+from dask.distributed import Client
+from dask.tokenize import tokenize
+from rich.console import Group
+from rich.live import Live
+from rich.progress import Progress
+
+from util._futures import DynamicAsCompleted, FutureLike
+
+try:
+    from coffea.util import coffea_console
+except ImportError:
+    from rich.console import Console
+
+    coffea_console = Console()
+
+
+_processing_sentinel = object()
+_final_merge_sentinel = object()
+
+
+# group of progress bars for dask/future executor
+def pbar_group(datasets: list[str]) -> tuple[Live, dict[tp.Any, Progress]]:
+    pbars = {_processing_sentinel: rich_bar()}
+    pbars.update({ds: rich_bar() for ds in datasets})
+    pbars[_final_merge_sentinel] = rich_bar()
+    return Live(Group(*pbars.values()), console=coffea_console), pbars
+
+
+Result: tp.TypeAlias = Accumulatable | BaseException
+
+
+@dataclasses.dataclass(frozen=True, slots=True)
+class Failure:
+    item: WorkItem
+    reason: BaseException
+
+
+class ReduceSchedulingError(RuntimeError): ...
+
+
+acc = partial(accumulate, accum=None)
+
+
+def failed_future(future: FutureLike) -> bool:
+    # if we return an exception as a value, consider it failed (see wrapped_process)
+    # we catch any exception, but no RuntimeError. Maybe the user wants to raise that?
+    return issubclass(future.type, BaseException) or future.status == "error"
+
+
+def wrapped_process(
+    process_func: tp.Callable[[ak.Array], Result],
+    workitem: WorkItem,
+    /,
+    *,
+    NanoEventsFactory_kwargs: dict[str, tp.Any] | None = None,
+) -> Result:
+    f = uproot.open(workitem.filename)
```

Comment:
I think it makes sense to have the equivalent of `NanoEventsFactory_kwargs` for the `uproot` call to set `array_cache`.

Reply:
ah wait, you may be able to pass uproot options like the array cache directly through `NanoEventsFactory_kwargs` and that one passes it down to each `tree.array(...)` call. I need to double check that.